readiness/controller/pkg/readiness/readiness_controller.go

```diff
@@ -34,14 +36,142 @@ type ReadinessReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
 func (r *ReadinessReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("readiness", req.NamespacedName)
+	readiness := &corev1alpha2.Readiness{}
+	err := r.Client.Get(ctx, req.NamespacedName, readiness)
+	if err != nil {
+		r.Log.Error(err, "Error while fetching readiness")
+		return ctrl.Result{}, err
+	}
 
-	// TODO(user): your logic here
+	updateRequired := false
+	oldStatus := readiness.Status
+
+	if len(readiness.Spec.Checks) == 0 {
+		if !oldStatus.Ready {
+			readiness.Status.Ready = true
+			readiness.Status.CheckStatus = []corev1alpha2.CheckStatus{}
+			updateRequired = true
+		}
+	} else {
+		allChecks := make(map[string][]int)
+		readiness.Status.CheckStatus = []corev1alpha2.CheckStatus{}
+		providersList := &corev1alpha2.ReadinessProviderList{}
+		err = r.Client.List(ctx, providersList)
+		if err != nil {
+			r.Log.Error(err, "Error while fetching readiness providers")
+			return ctrl.Result{}, err
+		}
+
+		for i, provider := range providersList.Items {
+			if _, ok := allChecks[provider.Spec.CheckRef]; !ok {
+				allChecks[provider.Spec.CheckRef] = make([]int, 0)
+			}
+
+			allChecks[provider.Spec.CheckRef] = append(allChecks[provider.Spec.CheckRef], i)
+		}
+
+		for _, check := range readiness.Spec.Checks {
+			checkStatus := corev1alpha2.CheckStatus{
+				Name:      check.Name,
+				Providers: make([]corev1alpha2.Provider, 0),
+				Ready:     false,
+			}
+
+			indices, ok := allChecks[check.Name]
+			if ok {
```

Comment:
Can be combined with L81
```suggestion
			if indices, ok := allChecks[check.Name]; ok {
```

Reply:
Done